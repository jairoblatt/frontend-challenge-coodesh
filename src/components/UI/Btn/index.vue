<template>
  <button v-on="$listeners" :class="classes">
    <spinner v-show="loading" size="21" />
    <slot v-show="!loading" />
  </button>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";

export default Vue.extend({
  name: "Btn",

  props: {
    primary: Boolean as PropType<boolean>,
    secondary: Boolean as PropType<boolean>,
    block: Boolean as PropType<boolean>,
    loading: Boolean as PropType<boolean>,
    small: Boolean as PropType<boolean>,
    outlined: Boolean as PropType<boolean>,
  },

  computed: {
    classes() {
      return {
        btn: true,
        "btn--primary": this.primary,
        "btn--secondary": this.secondary,
        "btn--block": this.block,
        "btn--small": this.small,
        "btn--outlined": this.outlined,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.btn {
  @apply rounded-md px-5 py-3 text-sm outline-none font-semibold text-white transition-colors;
}

.btn--primary {
  @apply bg-primary-base hover:bg-primary-light;
}

.btn--secondary {
  @apply bg-secondary-gray;
}

.btn--block {
  @apply w-full;
}

.btn:disabled {
  @apply bg-gray-300 text-gray-900 cursor-auto;
}

.btn--small {
  @apply px-7 py-2;
}

.btn--outlined {
  @apply border border-gray-100 text-light-onSurfacePrimary hover:border-black hover:text-black;
}
</style>
