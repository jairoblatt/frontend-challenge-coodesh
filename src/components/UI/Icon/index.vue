<template>
  <i v-if="iconSanitize" :class="`mdi ${iconSanitize}`" :style="styles"></i>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Icon",

  props: {
    size: {
      type: [String, Number],
      default: 14,
    } as PropOptions<string | number>,

    color: String as PropType<string>,

    icon: {
      type: String,
      default: "",
    } as PropOptions<string>,
  },

  computed: {
    iconSanitize(): string {
      return this.$slots.default?.[0]?.text || this.icon;
    },

    styles(): CSS.Properties {
      return {
        fontSize: this.size + "px",
        ...(this.color && { color: this.color }),
      };
    },
  },
});
</script>
