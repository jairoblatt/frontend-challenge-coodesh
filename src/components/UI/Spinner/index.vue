<template>
  <div class="spinner-container">
    <span class="spinner" :style="styles"></span>
  </div>
</template>
<script lang="ts">
import Vue, { PropOptions } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Spinner",

  props: {
    size: {
      type: [String, Number],
      default: "inherit",
    } as PropOptions<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        minHeight: `${this.size}px`,
        minWidth: `${this.size}px`,
      };
    },
  },
});
</script>
<style lang="postcss" scoped>
.spinner-container {
  @apply relative flex justify-center items-center min-h-full min-w-full;
}

.spinner {
  @apply pointer-events-none border-solid rounded-full  border-primary-base  p-1;

  border-width: 0.3rem;
  border-top-color: white;
  animation: spinner 0.5s linear infinite;
}

@keyframes spinner {
  100% {
    transform: rotate(360deg);
  }
}
</style>
