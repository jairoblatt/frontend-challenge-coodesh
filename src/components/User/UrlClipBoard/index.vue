<template>
  <div :class="classes">
    <input :value="locationHref" type="text" readonly @focus="copyToClipBoard" />
    <Btn data-testid="clipboard" @click="copyToClipBoard">
      <Icon :icon="iconClipBoard" />
    </Btn>
  </div>
</template>
<script lang="ts">
import Vue from "vue";
import { sleep } from "@/utils";

export default Vue.extend({
  data: () => ({
    copied: false,
  }),

  computed: {
    classes(): { [key: string]: boolean } {
      return {
        user__url: true,
        "user__url--copied": this.copied,
      };
    },

    locationHref(): string {
      return window.location.href;
    },

    iconClipBoard(): string {
      return this.copied ? "mdi-check" : "mdi-content-copy";
    },
  },

  methods: {
    async copyToClipBoard() {
      if (!this.copied) {
        await navigator.clipboard.writeText(this.locationHref);
        this.copied = true;

        await sleep(5000);
        this.copied = false;
      }
    },
  },
});
</script>
<style lang="postcss" scoped>
.user__url {
  @apply border border-light-onSurfacePrimary w-11/12 md:w-2/3 overflow-hidden border-opacity-40 border-solid rounded-sm mt-2 flex justify-between items-center;
}

.user__url input {
  @apply outline-none font-semibold px-2 w-full;
}
.user__url >>> button {
  @apply text-light-onSurfacePrimary border-l rounded-none border-light-onSurfacePrimary border-opacity-40 border-solid hover:bg-light-onSurfacePrimary hover:text-white;
}

.user__url--copied {
  @apply border-primary-base;
}

.user__url--copied >>> button {
  @apply border-primary-base hover:bg-primary-base bg-primary-base text-white;
}
</style>
