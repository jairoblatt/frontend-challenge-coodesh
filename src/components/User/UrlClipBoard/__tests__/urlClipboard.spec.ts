import { mount, Wrapper, MountOptions, createLocalVue } from "@vue/test-utils";
import UrlClipBoard from "../index.vue";
import Icon from "@/components/UI/Icon/index.vue";
import Btn from "@/components/UI/Btn/index.vue";
import Spinner from "@/components/UI/Spinner/index.vue";

const localVue = createLocalVue();
localVue.component("Icon", Icon);
localVue.component("Btn", Btn);
localVue.component("Spinner", Spinner);

type Instance = InstanceType<typeof UrlClipBoard>;
let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

const copyToClipBoardSpy = jest.spyOn((UrlClipBoard as any).options.methods, "copyToClipBoard");

describe("UrlClipBoard.vue ", () => {
  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(UrlClipBoard, {
        ...options,
        localVue,
      });
    };
  });

  it("Should mount the component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should have "user__url--copied" css class when the copied data it is true', async () => {
    const wrapper = mountFunction();
    wrapper.setData({
      copied: true,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.classes("user__url--copied")).toBe(true);
  });

  it("Should return the current icon", () => {
    const wrapper = mountFunction();

    expect((wrapper.vm as any).iconClipBoard).toBe("mdi-content-copy");

    wrapper.setData({
      copied: true,
    });

    expect((wrapper.vm as any).iconClipBoard).toBe("mdi-check");
  });
});
