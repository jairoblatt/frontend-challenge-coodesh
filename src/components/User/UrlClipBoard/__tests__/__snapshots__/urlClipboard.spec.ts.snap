// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UrlClipBoard.vue  Should mount the component 1`] = `
"<div class=\\"user__url\\"><input type=\\"text\\" readonly=\\"readonly\\"> <button class=\\"btn\\">
    <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 21px; min-width: 21px;\\"></span></div> <i class=\\"mdi mdi-content-copy\\" style=\\"font-size: 14px;\\"></i>
  </button></div>"
`;
